File: app/src/main/java/dcom/taneja/dhruv/floater/MainActivity.java
package dcom.taneja.dhruv.floater;

import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;

/**
 * Created by dhruvtaneja on 10/02/18.
 */

public class MainActivity extends AppCompatActivity {

    private static final int REQUEST_CODE = 123;
    private Intent mServiceIntent;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_main);

        findViewById(R.id.buttonStartFloater).setOnClickListener(view -> {
            if (!needsPermission()) {
                startFloater();
            } else {

                Intent intent = new Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION,
                        Uri.parse("package:" + getPackageName()));
                startActivityForResult(intent, REQUEST_CODE);
            }
        });

        findViewById(R.id.buttonStopFloater).setOnClickListener(view -> {
            if (mServiceIntent != null) {
                stopService(mServiceIntent);
            }
        });
    }

    private boolean needsPermission() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.M
                && !Settings.canDrawOverlays(getApplicationContext());
    }

    private void startFloater() {
        mServiceIntent = new Intent(this, FloaterService.class);
        startService(mServiceIntent);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == REQUEST_CODE && resultCode == RESULT_OK) {
            startFloater();
        }
        super.onActivityResult(requestCode, resultCode, data);
    }
}

File: app/src/main/java/dcom/taneja/dhruv/floater/MainActivity.kt
package dcom.taneja.dhruv.floater

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.provider.Settings
import dcom.taneja.dhruv.floater.service.FloaterService
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    private val requestCode = 123
    private var serviceIntent: Intent? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        buttonStartFloater.setOnClickListener {
            if (!needsPermission()) {
                startFloater()
            } else {
                val intent = Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION,
                        Uri.parse("package:$packageName"))
                startActivityForResult(intent, requestCode)
            }
        }

        buttonStopFloater.setOnClickListener {
            if (serviceIntent != null) {
                stopService(serviceIntent)
            }
        }
    }

    private fun startFloater() {
        serviceIntent = Intent(this, FloaterService::class.java)
        startService(serviceIntent)
    }

    @SuppressLint("ObsoleteSdkInt")
    private fun needsPermission() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.M
                    && !Settings.canDrawOverlays(applicationContext)

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == requestCode) {
            if (resultCode == Activity.RESULT_OK) {
                startFloater()
            }
        }
        super.onActivityResult(requestCode, resultCode, data)
    }
}
